adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -121,6 +121,121 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+   *                     upper bound
+   * @return An iterator of all T's keyed with the specific ReferenceRegion that the T represents in
+   *         the global sort
+   */
+  private def inferCorrectReferenceRegionsForPartition(
+    iter: Iterator[T], partitionMap: (ReferenceRegion, ReferenceRegion)): Iterator[((ReferenceRegion, T))] = {
+
+    // we put the records that map to multiple locations in this ListBuffer
+    // IMPORTANT: these are records that map to multiple locations on this
+    // partition. There should be very few of these
+    // we use this because we have to keep track of which ReferenceRegions
+    // have been used.
+    val duplicates = new ListBuffer[(Seq[ReferenceRegion], T)]()
+    // filter out ReferenceRegions that are not on this partition
+    iter.map(f => (getReferenceRegions(f).filter(g =>
+      g.compareTo(partitionMap._1) >= 0 && g.compareTo(partitionMap._2) <= 0).sorted, f))
+      .map(f => {
+        // the usual case: only one ReferenceRegion for the record
+        if (f._1.length == 1) {
+          (f._1.head, f._2)
+          // if there are multiple, we have to do a little extra work to make
+          // sure that we have the correct ReferenceRegion
+        } else {
+          val index = duplicates.indexOf(f)
+          // if we've already added this record to the duplicates list
+          if (index >= 0) {
+            // get the inferred value from the sorted list of ReferenceRegions
+            val inferredValue = (duplicates(index)._1.head, duplicates(index)._2)
+            // if there are more after this
+            if (duplicates(index)._1.length != 1) duplicates(index) = (duplicates(index)._1.drop(1), duplicates(index)._2)
+            inferredValue
+          } else {
+            val inferredValue = (f._1.head, f._2)
+            // add to the list of duplicates
+            duplicates += ((f._1.drop(1), f._2))
+            inferredValue
+          }
+        }
+      })
+  }
+
+  /**
+   * This is a function that splits each partition (iter) into lists keyed with the destination
+   * partition number. It preserves the order of the data received and thus is good for moving
+   * around sorted data. Each partition is given the same amount of data.
+   *
+   * @param iter The partition data. (k,v) where k is the index in the RDD and v is T
+   * @param average the average number of tuples on all partitions
+   * @return grouped lists keyed with the destination partition number. These lists maintain
+   *         their original order
+   */
+  private def getBalancedPartitionNumber(iter: Iterator[((ReferenceRegion, T), Long)], partitionMap: (ReferenceRegion, ReferenceRegion),
+                                         average: Double): Iterator[(Int, ((ReferenceRegion, ReferenceRegion), List[(Long, T)]))] = {
+
+    iter.map(_.swap).map(f => ((f._1 / average).toInt, f)).toList.groupBy(_._1).mapValues(f => f.map(_._2))
```

Comment:
Also, why are we converting to a list and grouping by key? That will force materialization of the whole iterator.